### DISCUSSION TITLE
Gmail's rollout of Trusted Types

### DISCUSSION THREAD

**Discussion by wegry** [InboxSDK/InboxSDK#1127] [General]
# Background

On 2024-02-12[^1], Gmail started to roll out Trusted Types support in Gmail. We had a user report issues around `Search.registerSearchSuggestionsProvider`[^2]. After checking our logs, we saw an assignment to `HTMLElement.innerHTML` failing[^3]. For users reporting, the error was:

```
TypeError: Failed to set the 'innerHTML' property on 'Element': This document requires 'TrustedHTML' assignment.`
```

As a preliminary measure, #1126 uses Trusted Types in browsers that support it[^4]. This fix is available in `@inboxsdk/core@2.1.30`. As of 3d13f4ec423fe68f689ab947de976c21c90380d3, we have 67 assignments to `HTMLElement.innerHTML` that are potentially affected by Gmail's update.[^5] 

# Going forward

There will presumably be more issues as Gmail continues to roll this change out. We'll be monitoring and [fixing](https://web.dev/articles/trusted-types#fix-the-violations) accordingly.

## Logging `securitypolicyviolation`

[MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/securitypolicyviolation_event) lays out how we might track offending CSP around `require-trusted-types-for` violations to track down how widespread the issue is.

## Using a default policy
Worst comes to worst, we might just turn off TrustedTypes in their entirety with a default Trusted Types policy.

```js
// https://web.dev/articles/trusted-types#fix-the-violations
if (window.trustedTypes && trustedTypes.createPolicy) { // Feature testing
  trustedTypes.createPolicy('default', {
    createHTML: (string, sink) => DOMPurify.sanitize(string, {RETURN_TRUSTED_TYPE: true})
  });
}
```

[^1]: https://workspaceupdates.googleblog.com/2024/01/extending-trusted-types-to-gmail.html
[^2]: https://github.com/InboxSDK/InboxSDK/blob/3d13f4ec423fe68f689ab947de976c21c90380d3/src/platform-implementation-js/namespaces/search.ts#L34
[^3]: https://github.com/InboxSDK/InboxSDK/blob/0a2f60e3e3450a0e425a320f32a05f3d6e7d39ac/src/common/html-to-text.ts#L5
[^4]: https://caniuse.com/trusted-types
[^5]: `$ rg "\.innerHTML =" -g '!examples/' -g '!**/*.test.ts' -g '!__tests__/' --stats`

**Comment by Macil** [comment]
As of v2.1.30 (Feb 15), the InboxSDK is compatible with Gmail's use of Trusted Types. In versions before this, the `Search.registerSearchSuggestionsProvider` function will fail to add search suggestions to Gmail.

Extensions using InboxSDK v2.1.30 should have no issues with Gmail's use of Trusted Types unless they themselves are specifically using scripts in the "main"/"page" world instead of the default content script isolated world. Content scripts within the default isolated world are unaffected by Gmail's use of Trusted Types. See https://developer.chrome.com/docs/extensions/develop/concepts/content-scripts#isolated_world for an explanation of script worlds. Here are some ways to tell if you may be affected:
* You have a content script declared in your manifest.json with the `"world": "MAIN"` option set.
* You have a background page / service worker with a call to `chrome.scripting.executeScript` using the `world: "MAIN"` option with a file other than the InboxSDK-provided "pageWorld.js".
* You have a call to `document.createElement("script")` and append the script to the page within Gmail. (Important notice unrelated to Trusted Types: This will not work in MV3, so you must update your extension to remove this before Chrome requires extensions to use MV3.)
